pkg/object/meshcontroller/service/service.go

```diff
@@ -942,3 +943,90 @@ func (s *Service) PutTrafficTarget(tt *spec.TrafficTarget) {
 		api.ClusterPanic(err)
 	}
 }
+
+// PutServiceCanarySpec updates the service canary spec.
+func (s *Service) PutServiceCanarySpec(serviceCanarySpec *spec.ServiceCanary) {
+	buff, err := yaml.Marshal(serviceCanarySpec)
+	if err != nil {
+		panic(fmt.Errorf("BUG: marshal %#v to yaml failed: %v", serviceCanarySpec, err))
+	}
+
+	err = s.store.Put(layout.ServiceCanaryKey(serviceCanarySpec.Name), string(buff))
+	if err != nil {
+		api.ClusterPanic(err)
+	}
+}
+
+// GetServiceCanary gets the service canary.
+func (s *Service) GetServiceCanary(serviceCanaryName string) *spec.ServiceCanary {
+	serviceCanary, _ := s.GetServiceCanaryWithInfo(serviceCanaryName)
+	return serviceCanary
+}
+
+// GetServiceCanaryWithInfo gets the service canary with raw info.
+func (s *Service) GetServiceCanaryWithInfo(serviceCanaryName string) (*spec.ServiceCanary, *mvccpb.KeyValue) {
```

Comment:
this function is only called in `GetServiceCanary` and `WithInfo` is not good name, I propose to remove this function and only keep `GetServiceCanary`

Reply:
Updated